### DISCUSSION TITLE
Tracks playing out of Sync

### DISCUSSION THREAD

**Discussion by phonographic** [katspaugh/wavesurfer.js#2868] [Q&A]
I am working on a plugin that plays 2 tracks at the same time. The waveforms are overlaid. I have 3 buttons that allow me to switch between track1 only (Track2 does not display and is muted / volume 0) Track 1&2 (both tracks display) Track2 (Only track 2 displays and is muted / volume 0). Everything is working fine but when I toggle between these states, one of the tracks will go out of sync. I have tried a few things but no success. Here is the js code. Can anyone help?

`class AudioWaveJS extends elementorModules.frontend.handlers.Base {
    
    getDefaultSettings() {
        return {
            isPlaying:false,
            isAlone:false,
            isOther:false,
            isFull:true,
            masterClock:0,
            selectors: {
                audioMixerContainer:'.audio-mixer-container',
                playButton:'.play-button',
                aloneButton:'.alone-button',
                fullButton:'.full-button',
                otherButton:'.other-button',
                waveContainer: '.waveform-container-parent',
                waveContainer1:'.waveform-container-1',
                timelineContainer1:'.timeline-container-1',
                waveContainer2:'.waveform-container-2',
                composerContainer:'.composer',
                titleContainer:'.song-title',
                progressbarContainer: '.progress-bar-container',
                buttonsContainer:'.buttons-container',
            },
        };
    }
    
    getDefaultElements() {
        const selectors = this.getSettings( 'selectors' );
        
        return {
            $waveContainer: this.$element.find( selectors.waveContainer ),
            $playButton: this.$element.find( selectors.playButton ),
            $aloneButton: this.$element.find( selectors.aloneButton ),
            $fullButton: this.$element.find( selectors.fullButton ),
            $otherButton: this.$element.find( selectors.otherButton ),
            $waveContainer1: this.$element.find( selectors.waveContainer1 ),
            $timelineContainer1: this.$element.find( selectors.timelineContainer1 ),
            $waveContainer2: this.$element.find( selectors.waveContainer2 ),
            $secondSelector: this.$element.find( selectors.secondSelector ),
        };
    }
    bindEvents() {
        this.elements.$playButton.on( 'click', () => this.togglePlayback() );
        this.elements.$aloneButton.on( 'click', () => this.toggleAlone() );
        this.elements.$fullButton.on( 'click', () => this.toggleFull() );
        this.elements.$otherButton.on( 'click', () => this.toggleOther() );
    }
    onInit() {
		if ( this.isActive( this.getSettings() ) ) {
			elementorModules.ViewModule.prototype.onInit.apply( this, arguments );
		}
		let ele_cont='.elementor-element-'+this.getID();
		 this.audio1 = WaveSurfer.create({
            container: this.getDefaultElements().$waveContainer1[0],
            waveColor: "#4F4F4F",
            progressColor: "#FFF",
            //cursorColor: "#FF4081",
            cursorColor: "transparent",
            barWidth: 2,
            barRadius: 0,
            barHeight: 1,
            height: 60,
            responsive: true,
            url:this.getElementSettings('track_url_1').url,
            
          });
        if(this.getElementSettings('track_url_2')!==undefined){
            this.audio2 = WaveSurfer.create({
                container: this.getDefaultElements().$waveContainer2[0],
                waveColor: "#4F4F4F",
                progressColor: "#FFF",
                cursorColor: "transparent",
                barWidth: 2,
                barRadius: 0,
                barHeight: 1,
                height: 60,
                responsive: true,

              });
              // Load tracks
            //this.audio1.load(this.getElementSettings('track_url_1').url);
            this.audio2.load(this.getElementSettings('track_url_2').url);
            this.audio2.on('interaction', () => {
             
              (this.audio1.setTime(this.audio2.getCurrentTime()));
            });
            this.audio2.on('drag',(r)=>{
                this.audio1.setTime(this.audio1.getDuration()*r);
            })
            this.audio1.on('interaction', () => {
             
              (this.audio2.setTime(this.audio1.getCurrentTime()));
            });
            this.audio2.on('finish',()=>{
                (this.audio2.setTime(0));
                (this.audio1.setTime(0));
                this.togglePlayback();
            });
        }else{
            this.audio1.on('finish',()=>{
                
                (this.audio1.setTime(0));
                this.togglePlayback();
            });
        }
		 
            
            
            this.initAudio();
	}
	
	initAudio(){
	    let ele_cont='.elementor-element-'+this.getID();
	    let that =this;
	    
        
            
            
        //that.audio1.addPlugin(WaveSurfer.timeline.create({container: timelineContainer1,})).initPlugin('timeline');
         //that.audio1.registerPlugin(WaveSurfer.Timeline.create({insertPosition:'beforebegin'}));
        
       /* this.audio2.on("ready", function () {
           
        
            that.audio2.registerPlugin(WaveSurfer.Timeline.create());
          });*/
	}
	
	togglePlayback() {
	    let isPlaying=this.getSettings( 'isPlaying' );
	    this.setSettings({
          isPlaying: !isPlaying,
        });
        if (isPlaying) {
          this.audio1.pause();
          (this.getElementSettings('multitrack')!=undefined) && this.audio2.pause();
          window.AudioWaveJS.playing = false;
          window.AudioWaveJS.instance=null;
          window.AudioWaveJS.elementId = null;
          this.elements.$playButton.removeClass('active'); // Play triangle symbol
          
        } else {
          this.audio1.play();
          (this.getElementSettings('multitrack')!=undefined) && this.audio2.play();
          if(window.AudioWaveJS.elementId != this.getID()){
              if(window.AudioWaveJS.elementId !== null){
                  window.AudioWaveJS.instance.setSettings({
                      isPlaying: false,
                    });
                  window.AudioWaveJS.instance.audio1.pause();
                  (window.AudioWaveJS.instance.audio2!=undefined) && window.AudioWaveJS.instance.audio2.pause();
                  window.AudioWaveJS.instance.elements.$playButton.removeClass('active');
              }
              //console.log(window.AudioWaveJS.instance);
          }
          window.AudioWaveJS.playing = true;
          window.AudioWaveJS.instance=this;
          window.AudioWaveJS.elementId =this.getID();
          this.elements.$playButton.addClass('active') // Pause symbol
          
        }
        
        
     }
    toggleAlone() {
        let isAlone=this.getSettings( 'isAlone' );
        if(isAlone) return false;
        this.setSettings({
          isAlone: !isAlone,
          isOther:false,
          isFull:false,
        })
        
        if (!isAlone) {
          if(this.getOS()=='iOS'){
              this.audio1.setMuted(0);
              (this.getElementSettings('multitrack')!=undefined) && this.audio2.setMuted(1);
          }else{
            this.audio1.setVolume(1);
            (this.getElementSettings('multitrack')!=undefined) && this.audio2.setVolume(0);
          }
          //
          this.elements.$aloneButton.addClass("active");
          this.elements.$fullButton.removeClass("active");
          this.elements.$otherButton.removeClass("active");
        } 
    
        this.toggleWaveformContainers();
    }
    toggleFull() {
        let isAlone=this.getSettings( 'isAlone' );
        let isOther=this.getSettings( 'isOther' );
        let isFull=this.getSettings( 'isFull' );
        if(isFull) return false;
        this.setSettings({
          isAlone: false,
          isOther:false,
          isFull:!isFull,
        });
        if (!isFull) {
          if(this.getOS()=='iOS'){
              this.audio1.setMuted(0);
              (this.getElementSettings('multitrack')!=undefined) && this.audio2.setMuted(0);
          }else{
             this.audio1.setVolume(1);
            (this.getElementSettings('multitrack')!=undefined) && this.audio2.setVolume(1);
          }
          
          this.elements.$aloneButton.removeClass("active");
          this.elements.$fullButton.addClass("active");
          this.elements.$otherButton.removeClass("active");
        } 
        
       
        this.toggleWaveformContainers();
    }
    toggleOther() {
        let isAlone=this.getSettings( 'isAlone' );
        let isOther=this.getSettings( 'isOther' );
        let isFull=this.getSettings( 'isFull' );
        if(isOther) return false;
        this.setSettings({
          isAlone: false,
          isOther:!isOther,
          isFull:false,
        });
        if (!isOther) {
          if(this.getOS()=='iOS'){
              this.audio1.setMuted(1);
              (this.getElementSettings('multitrack')!=undefined) && this.audio2.setMuted(0);
          }else{
              this.audio1.setVolume(0);
              (this.getElementSettings('multitrack')!=undefined) && this.audio2.setVolume(1);
          }
          
          
          this.elements.$aloneButton.removeClass("active");
          this.elements.$fullButton.removeClass("active");
          this.elements.$otherButton.addClass("active");
        }
        
        
        this.toggleWaveformContainers();
    }
    getOS() {
      var userAgent = window.navigator.userAgent,
          platform = window.navigator?.userAgentData?.platform || window.navigator.platform,
          macosPlatforms = ['Macintosh', 'MacIntel', 'MacPPC', 'Mac68K'],
          windowsPlatforms = ['Win32', 'Win64', 'Windows', 'WinCE'],
          iosPlatforms = ['iPhone', 'iPad', 'iPod'],
          os = null;
    
      if (macosPlatforms.indexOf(platform) !== -1) {
        os = 'Mac OS';
      } else if (iosPlatforms.indexOf(platform) !== -1) {
        os = 'iOS';
      } else if (windowsPlatforms.indexOf(platform) !== -1) {
        os = 'Windows';
      } else if (/Android/.test(userAgent)) {
        os = 'Android';
      } else if (/Linux/.test(platform)) {
        os = 'Linux';
      }
    
      return os;
    }
    toggleWaveformContainers() {
        let isAlone=this.getSettings( 'isAlone' );
        let isOther=this.getSettings( 'isOther' );
        
        const $waveformContainer1 = this.getDefaultElements().$waveContainer1;
        const $waveformContainer2 = this.getDefaultElements().$waveContainer2;
    
        if (isAlone) {
          $waveformContainer1.addClass("full-display");
          $waveformContainer1.removeClass("hide-display");
          $waveformContainer2.removeClass("full-display");
          $waveformContainer2.addClass("hide-display");
        } else if (isOther) {
          $waveformContainer1.removeClass("full-display");
          $waveformContainer1.addClass("hide-display");
          $waveformContainer2.addClass("full-display");
          $waveformContainer2.removeClass("hide-display");
        } else {
          $waveformContainer1.addClass("full-display").removeClass('hide-display');
          
          $waveformContainer2.addClass("full-display").removeClass('hide-display');
        }
    }
}

jQuery( window ).on( 'elementor/frontend/init', () => {

    window.AudioWaveJS ={
        playing:false,
        elementId:null,
        instance:null,
    };
    
  elementorFrontend.elementsHandler.attachHandler( 'audio-mixer-of-love', AudioWaveJS );
  
  
 } );

`

**Comment by katspaugh** [comment]
You might check how [Multitrack](https://github.com/katspaugh/wavesurfer-multitrack) does it. Basically subscribe to the timeupdate event and call setTime on the other track if it’s time differs too much from the first one.
